runtime:vim: make K / &keywordprg smarter

Konfekt opened the pull request:
Let keywordprg in vim.vim to handle context-sensitive help calls. Detect the syntax group of the word under the cursor.

See https://github.com/vim/vim/issues/16677

Konfekt committed:
runtime:vim: make K / &keywordprg smarter

Let keywordprg in vim.vim to handle context-sensitive help calls.
Detect the syntax group of the word under the cursor.

See https://github.com/vim/vim/issues/16677

diff --git a/runtime/ftplugin/vim.vim b/runtime/ftplugin/vim.vim:
-    setl fo< isk< com< tw< commentstring< include< define< keywordprg<
+    setl fo< isk< com< tw< commentstring< include< define<
-" Distinguish between commands, options and functions.
-function! s:Help(args)
-  silent! let syn_name = synIDattr(synID(line('.'), col('.'), 1), 'name')
-
-  if syn_name =~# 'vimCommand'
-    execute 'help' a:args..':'
-  elseif syn_name =~# 'vimOption'
-    execute 'help' "'"..a:args.."'"
-  elseif syn_name =~# 'vimFunc'
-    execute 'help' a:args..'()'
-  else
-    execute 'help' a:args
-  endif
-endfunction
-command! -buffer -nargs=1 VimKeywordPrg :call s:Help(<q-args>)
-setlocal keywordprg=:VimKeywordPrg
+setlocal keywordprg=:help

AndrewRadev commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
+  silent! let syn_name = synIDattr(synID(line('.'), col('.'), 1), 'name')
+
+  if syn_name =~# 'vimCommand'
+    execute 'help' a:args..':'
```
```suggestion
    execute 'help :'..a:args
```

Konfekt replied:
Thank you @AndrewRadev . Now it show's that I just tested the distinction between functions and variables.

benknoble commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
+function! s:Help(args)
+  silent! let syn_name = synIDattr(synID(line('.'), col('.'), 1), 'name')
+
+  if syn_name =~# 'vimCommand'
```
Since syntax may not be on, it seems prudent to perform the (probably cheaper) `&syntax` check first. Something like

```
if syntax on | execute 'help' a:args | return | endif

" Rest of what we have
```

chrisbra replied:
I suppose you mean `if  g:syntax_on`

benknoble replied:
Right, pseudocode ;) should have indicated that better

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
 " Use :help to lookup the keyword under the cursor with K.
-setlocal keywordprg=:help
+" Distinguish between commands, options and functions.
+function! s:Help(args)
```
Please also add:
```
  if exists("*" .. expand("<SID>") .. "Help")
      finish
   endif
```

Konfekt replied:
I am not sure. Doens't finish come a bit early here, since `s:help` will be defined on the first load of this file, and then the following settings no longer applied in other buffers?

Konfekt replied:
I guess we can go without finishing here

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
+  if syn_name =~# 'vimCommand'
+    execute 'help :'..a:args
+  elseif syn_name =~# 'vimOption'
+    execute 'help' "'"..a:args.."'"
```
```suggestion
    execute "help '"..a:args.."'"
```

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
+  elseif syn_name =~# 'vimOption'
+    execute 'help' "'"..a:args.."'"
+  elseif syn_name =~# 'vimFunc'
+    execute 'help' a:args..'()'
```
inconsistent concat

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +48,22 @@ setlocal fo-=t fo+=croql
+  elseif syn_name =~# 'vimFunc'
+    execute 'help' a:args..'()'
+  else
+    execute 'help' a:args
```
inconsistent concat

Konfekt committed:
apply suggestions by @chrisbra, @benknoble, @AndrewRadev and @Gary

Konfekt committed:
driveby maintenance

Konfekt committed:
stylistic amendments

Konfekt committed:
play it safe

Konfekt committed:
improve using @tpope 's scriptease [0] as suggested by @dkearn

Tested with:

execute('somecommand') " --> opens :execute instead of execute() help
echo match("testing", "ing") " --> opens :match instead of match() help
echo escape('c:\program files\vim', ' \') " --> opens escape instead of escape()
setlocal complete " --> opens complete() instead of 'complete'

[0]: https://github.com/tpope/vim-scriptease/blob/cdb5981d47ac98221a408ae2e7cae66524d9e872/autoload/scriptease.vim#L738-L766

Konfekt committed:
checking for command by syntax highlighting shouldn't hurt

Konfekt committed:
simplify and fix typo

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -48,7 +49,49 @@ setlocal fo-=t fo+=croql
+    let topic = a:topic
+
+    if get(g:, 'syntax_on', 0)
+      silent! let syn = synIDattr(synID(line('.'), col('.'), 1), 'name')
```
why the silent!?

Konfekt replied:
It will return an empty string, falling back to `:help` if it fails; not worth bothering long-term users with potential error messages

chrisbra replied:
and why would this fail when syntax highlighting is enabled?

Konfekt replied:
It's just a precautionary measure to exclude unpredicted failures

dkearns replied:
I can't see how that can fail.

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -101,27 +144,27 @@ if exists("loaded_matchit")
+        \ '\<aug\%[roup]\s\+\%(END\>\)\@!\S:\<aug\%[roup]\s\+END\>,' ..
+        \ '\<class\>:\<endclass\>,' ..
+        \ '\<interface\>:\<endinterface\>,' ..
+        \ '\<enum\>:\<endenum\>'
```
please don't change the indentation.

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -101,27 +144,27 @@ if exists("loaded_matchit")
   " Also ignore here-doc and dictionary keys (vimVar).
   let b:match_skip = 'synIDattr(synID(line("."), col("."), 1), "name")
-	\ =~? "comment\\|string\\|vimSynReg\\|vimSet\\|vimLetHereDoc\\|vimVar"'
+        \ =~? "comment\\|string\\|vimSynReg\\|vimSet\\|vimLetHereDoc\\|vimVar"'
```
same here

chrisbra commented on runtime/ftplugin/vim.vim:
```diff
@@ -101,27 +144,27 @@ if exists("loaded_matchit")
 
 " removed this, because 'cpoptions' is a global option.
-" setlocal cpo+=M		" makes \%( match \)
+" setlocal cpo+=M               " makes \%( match \)
```
same here

Konfekt replied:
Why the mix of tabs and spaces? This was not intentional, just  a recurrent headache when contributing.

dkearns replied:
The runtime files were, for the longest time, supposed to be `ts=8 noet`.  That's probably been relaxed.

I think we generally prefer dedicated whitespace fix commits if making those changes but I don't really think they're warranted.  How about adding the modeline?

Please add yourself as a contributor, if you'd like.

Konfekt replied:
Which modeline, the one setting`ts=8 noet` or `sw=2 et` ?

Konfekt committed:
white space adjustments

Konfekt committed:
apply @dkean 's suggestions
